src/mapmvt.c

```diff
@@ -543,12 +543,34 @@ int msMVTWriteTile(mapObj *map, int sendheaders) {
     features_size = FEATURES_INCREMENT_SIZE;
 
     msInitShape(&shape);
-    while ((status = msLayerNextShape(layer, &shape)) == MS_SUCCESS) {
+    int nshapes;
+    nshapes = layer->resultcache ? layer->resultcache->numresults
+                                 : msLayerGetShapeCount(layer, rect, NULL);
+    for (i = 0; i < nshapes; i++) {
+      status = layer->resultcache
+                   ? msLayerGetShape(layer, &shape,
+                                     &(layer->resultcache->results[i]))
+                   : msLayerNextShape(layer, &shape);
+      if (status == MS_FAILURE)
+        goto feature_cleanup;
 
       if (layer->numclasses > 0) {
-        shape.classindex = msShapeGetClass(
-            layer, map, &shape, NULL, -1); /* Perform classification, and some
-                                              annotation related magic. */
+        int nclasses = 0;
+        int *classgroup = NULL;
+
+        if (layer->classgroup && layer->numclasses > 0)
```

Comment:
There might be a performance implication in doing msAllocateValidClassGroups / msFree within the loop. Looking at msQueryByFilter(), the call to msAllocateValidClassGroups is done outside the loop

Reply:
Yes of course. I don't know why I didn't see it...